File: app/schemas/procurement.py
"""采购需求相关 Schema"""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ProcurementCreate(BaseModel):
    """创建采购需求请求"""

    title: str = Field(..., min_length=2, max_length=200, description="采购标题")
    description: Optional[str] = Field(None, description="需求详细描述")
    category: str = Field(..., min_length=2, max_length=100, description="采购品类")
    budget_min: Optional[float] = Field(None, ge=0, description="预算下限")
    budget_max: Optional[float] = Field(None, ge=0, description="预算上限")
    deadline: Optional[str] = Field(None, description="采购截止日期")


class ProcurementUpdate(BaseModel):
    """更新采购需求请求"""

    title: Optional[str] = Field(None, min_length=2, max_length=200)
    description: Optional[str] = None
    category: Optional[str] = None
    budget_min: Optional[float] = None
    budget_max: Optional[float] = None
    deadline: Optional[str] = None
    status: Optional[str] = None


class ProcurementResponse(BaseModel):
    """采购需求响应"""

    id: int
    visitor_id: int
    visitor_name: Optional[str] = None
    title: str
    description: Optional[str] = None
    category: str
    budget_min: Optional[float] = None
    budget_max: Optional[float] = None
    deadline: Optional[datetime] = None
    status: str
    match_count: int = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MatchCreate(BaseModel):
    """创建采购匹配请求"""

    procurement_id: int = Field(..., description="采购需求ID")
    message: Optional[str] = Field(None, description="展商留言/报价说明")
    quoted_price: Optional[float] = Field(None, ge=0, description="报价")


class MatchResponse(BaseModel):
    """采购匹配响应"""

    id: int
    procurement_id: int
    exhibitor_id: int
    exhibitor_name: Optional[str] = None
    company_name: Optional[str] = None
    message: Optional[str] = None
    quoted_price: Optional[float] = None
    is_accepted: Optional[bool] = None
    created_at: datetime

    class Config:
        from_attributes = True

File: expo-hub-frontend/src/api/procurement.ts
import http from './index'
import type { ProductItem } from '@/api/product'
import type { ListResp } from './paged'
import { normList } from './paged'

export interface ProcurementItem {
  id: number
  visitor_id: number
  visitor_username?: string
  title: string
  description?: string
  category: string
  // 撮合平台后端刻意不暴露价格/数量（置 null）
  budget_min?: number | null
  budget_max?: number | null
  quantity?: number | null
  unit?: string | null
  deadline?: string
  status: string
  match_count: number
  created_at: string
  updated_at: string
}

export interface ProcurementMatch {
  id: number
  procurement_id: number
  exhibitor_id: number
  exhibitor_username?: string
  exhibitor_company?: string
  message?: string
  quoted_price?: number
  is_accepted?: boolean
  created_at: string
}

export interface CreateProcurementParams {
  title: string
  description?: string
  category: string
  budget_min?: number
  budget_max?: number
  deadline?: string
}

// Recommended product from procurement matching
export interface MatchProductItem extends ProductItem {
  match_score?: number
  match_reasons?: string[]
}

export const procurementApi = {
  async getList(params?: { category?: string; status?: string; keyword?: string; search?: string; page?: number; page_size?: number }): Promise<ListResp<ProcurementItem>> {
    const res: any = await http.get('/procurements', { params })
    return normList<ProcurementItem>(res)
  },
  getDetail(id: number): Promise<ProcurementItem> {
    return http.get(`/procurements/${id}`) as Promise<ProcurementItem>
  },
  create(data: CreateProcurementParams): Promise<ProcurementItem> {
    return http.post('/procurements', data) as Promise<ProcurementItem>
  },
  update(id: number, data: Partial<CreateProcurementParams>): Promise<ProcurementItem> {
    return http.put(`/procurements/${id}`, data) as Promise<ProcurementItem>
  },
  cancel(id: number): Promise<ProcurementItem> {
    return http.post(`/procurements/${id}/cancel`) as Promise<ProcurementItem>
  },
  // 匹配相关
  async getMatches(procurementId: number): Promise<ListResp<ProcurementMatch>> {
    const res: any = await http.get(`/procurements/${procurementId}/matches`)
    return normList<ProcurementMatch>(res)
  },
  createMatch(procurementId: number, data: { message?: string; quoted_price?: number }): Promise<ProcurementMatch> {
    return http.post(`/procurements/${procurementId}/matches`, data) as Promise<ProcurementMatch>
  },
  acceptMatch(procurementId: number, matchId: number): Promise<ProcurementMatch> {
    return http.post(`/procurements/${procurementId}/matches/${matchId}/accept`) as Promise<ProcurementMatch>
  },
  // 游客自己的采购需求
  async getMyProcurements(params?: { status?: string; page?: number; page_size?: number }): Promise<ListResp<ProcurementItem>> {
    const res: any = await http.get('/procurements/my', { params })
    return normList<ProcurementItem>(res)
  },
  async getRecommendations(id: number): Promise<ListResp<MatchProductItem>> {
    const res: any = await http.get('/procurements/' + id + '/recommendations')
    return normList<MatchProductItem>(res)
  }
}

// Type alias for views that import `type Procurement`
export type Procurement = ProcurementItem
